diesel_derives/src/lib.rs

```diff
@@ -1627,3 +1631,151 @@ pub fn table_proc(input: TokenStream) -> TokenStream {
 pub fn derive_multiconnection(input: TokenStream) -> TokenStream {
     multiconnection::derive(syn::parse_macro_input!(input)).into()
 }
+
+/// Automatically annotates return type of a query fragment function
+///
+/// This may be useful when factoring out common query fragments into functions.
+/// If not using this, it would typically involve explicitly writing the full
+/// type of the query fragment function, which depending on the length of said
+/// query fragment can be quite difficult (especially to maintain) and verbose.
+///
+/// # Example
+///
+/// ```rust
+/// # extern crate diesel;
+/// # include!("../../diesel/src/doctest_setup.rs");
+/// # use schema::{users, posts};
+/// use diesel::dsl;
+///
+/// # fn main() {
+/// #     run_test().unwrap();
+/// # }
+/// #
+/// # fn run_test() -> QueryResult<()> {
+/// #     let conn = &mut establish_connection();
+/// #
+/// #[dsl::auto_type]
+/// fn user_has_post() -> _ {
+///     dsl::exists(posts::table.filter(posts::user_id.eq(users::id)))
+/// }
+///
+/// let users_with_posts: Vec<String> = users::table
+///     .filter(user_has_post())
+///     .select(users::name)
+///     .load(conn)?;
+///
+/// assert_eq!(
+///     &["Sean", "Tess"] as &[_],
+///     users_with_posts
+///         .iter()
+///         .map(|s| s.as_str())
+///         .collect::<Vec<_>>()
+/// );
+/// #     Ok(())
+/// # }
+/// ```
+///
+/// # Advanced usage
+///
+/// By default, the macro will:
+///  - Generate a type alias for the return type of the function, named the
+///    exact same way as the function itself.
+///  - Assume that functions, unless otherwise annotated, have a type alias for
+///    their return type available at the same path as the function itself
+///    (including case). (e.g. for the `dsl::not(x)` call, it expects that there
+///    is a `dsl::not<X>` type alias available)
+///  - Assume that methods, unless otherwise annotated, have a type alias
+///    available as `diesel::dsl::PascalCaseOfMethodName` (e.g. for the
+///    `x.and(y)` call, it expects that there is a `diesel::dsl::And<X, Y>` type
+///    alias available)
+///
+/// The defaults can be changed by passing the following attributes to the
+/// macro:
+/// - `#[auto_type(no_type_alias)]` to disable the generation of the type alias.
+/// - `#[auto_type(dsl_path = "path::to::dsl")]` to change the path where the
+///   macro will look for type aliases for methods.
```

Comment:
We might want to be explicit here that this is "required" if you mix your own custom query dsl extensions with diesel types. Maybe we should just document a pattern for that? Something like create your own `dsl` module somewhere as follows:

```rust
mod dsl {
    /// export all of diesel dsl
    pub use diesel::dsl::*;

    /// Export your extension types here
    pub use crate::your_extension::dsl::YourType;
}

```